hail/src/main/scala/is/hail/expr/ir/InferPType.scala

```diff
@@ -426,14 +427,43 @@ object InferPType {
 
         branchType.setRequired(branchType.required && cond.pType2.required)
       }
-
       case Coalesce(values) =>
         getNestedElementPTypes(values.map( theIR => {
           InferPType(theIR, env)
           theIR._pType2
         }))
       case In(_, pType: PType) => pType
-      case _: ReadPartition | _: MakeStream => throw new Exception("Node not supported")
+      case x if x.typ == TVoid => {
+        x.children.map(c => InferPType(c.asInstanceOf[IR], env))
+        PVoid
+      }
+      case x@ResultOp(_, _) =>  PType.canonical(x.typ)
+      case CollectDistributedArray(contextsIR, globalsIR, contextsName, globalsName, bodyIR) => {
+        InferPType(contextsIR, env)
+        InferPType(globalsIR, env)
+        InferPType(bodyIR, env.bind(contextsName -> contextsIR._pType2, globalsName -> globalsIR._pType2))
+
+        PCanonicalArray(bodyIR._pType2, contextsIR._pType2.required)
+      }
+      case ReadPartition(rowIR, codecSpec, rowType) => {
+        InferPType(rowIR, env)
+
+        val child = codecSpec.buildDecoder(rowType)._1
+
+        PStream(child, child.required)
+      }
+      case MakeStream(irs, t) => {
+        if (irs.length == 0) {
+          PType.canonical(t, true).deepInnerRequired(true)
+        }
+
+        PStream(getNestedElementPTypes(irs.map( theIR => {
+          InferPType(theIR, env)
+          theIR._pType2
+        })), true)
+      }
+      case _:AggLet | _:ArrayAgg | _:ArrayAggScan | _:RunAgg | _:RunAggScan | _:NDArrayAgg | _:AggFilter | _:AggExplode |
```

Comment:
these would technically need to infer their children too, but they're going away soon so let's ignore that.

Reply:
yeah thought this was fine for that reason